### ISSUE TITLE
Halo exchanges necessitate copying the entire Dat back from device

### ISSUE THREAD

**Issue by wence-** [OP2/PyOP2#232]
At the moment, the halo exchange for a Dat when we're running on device looks like this:

``` python
    def halo_exchange_begin(self):
        if self.dataset.halo is None:
            return
        maybe_setflags(self._data, write=True)
        self._from_device()
        super(Dat, self).halo_exchange_begin()

    def halo_exchange_end(self):
        if self.dataset.halo is None:
            return
        maybe_setflags(self._data, write=True)
        super(Dat, self).halo_exchange_end()
        if self.state in [DeviceDataMixin.DEVICE,
                          DeviceDataMixin.BOTH]:
            self._halo_to_device()
            self.state = DeviceDataMixin.DEVICE

    def _halo_to_device(self):
        _lim = self.dataset.size * self.dataset.cdim
        self._device_data.ravel()[_lim:].set(self._data[self.dataset.size:])
```

Note that `_from_device` copies the entire Dat down to the host.  I think this can be improved by only (as in the `_halo_to_device` function) copying the necessary bits.  However, it's possible that there is a problem with multi-indexing device arrays.  If so, it may be simplest to always maintain the arrays on the device as logically 1D (they're 1D under the hood anyway), which would simplify indexing.

**Comment by wence-** [comment]
No longer relevant after #507.
